### ISSUE TITLE
React Native support

### ISSUE THREAD

**Issue by nickpoorman** [plouc/nivo#373]
Any plans to have this work with React Native?

Has anyone tried getting this working with a WebView? I've seen someone hack together a WebView for highcharts.  https://github.com/TradingPal/react-native-highcharts/blob/master/App/index.js

**Comment by marnett-git** [comment]
Has there been any progress on this? 
Did you manage to find a solution @nickpoorman?
I have found Victory has react-native support, but I prefer this library much more.

**Comment by nickpoorman** [comment]
@marnett-git - no but I stopped looking.

**Comment by jbevarts** [comment]
Would like to react-native support, please refactor the div's

**Comment by arjunk** [comment]
Yes, react-native support would be really good. I like the visualisation of this library. We have a product that has both a web application using react js and native android application using react-native. We would like to use a library that can provide visualisations with a similar look and feel for both the applications.

**Comment by bkniffler** [comment]
Victory is doing that @arjunk. But tbh, nivo charts looks so gorgeous, its hard to even consider any other charting library. Would totally go for nivo over victory, except if universal apps are a big deal.

**Comment by Amacelia** [comment]
Anyone got this working with react-native webview?

**Comment by liamjones** [comment]
I wonder how easily the SVG versions would work under react-native-svg? I can imagine the interactions might be an issue...

**Comment by plouc** [comment]
@liamjones, I have no experience with it, but would definitely lower the effort.

**Comment by gthom9** [comment]
+1 Would love to see React Native support.

**Comment by raufabr** [comment]
+1 Would be great to use Nivo with React Native

**Comment by scinfu** [comment]
+1

**Comment by shabnam701** [comment]
+1

**Comment by liamjones** [comment]
I would still like this. If someone has time to work on this, it might be worth looking at how [victory-native](https://github.com/FormidableLabs/victory-native) does this for victory, it might offer some tips.

**Comment by esthor** [comment]
> I would still like this. If someone has time to work on this, it might be worth looking at how [victory-native](https://github.com/FormidableLabs/victory-native) does this for victory, it might offer some tips.

I would also like this, and agree it might be worth looking at successful react native graph libraries! But I wouldn't suggest victory-native as a good modern example. 😞  There are lots of known performance bottlenecks (re-renders and frame drops; 'touch -> display overlays' and 'panning -> display discreet values' are notoriously difficult to get working in a performant way).

There seems to be a more modern architectural approach that avoids those problems today: [reanimated](https://github.com/software-mansion/react-native-reanimated) + [gesture handler](https://github.com/software-mansion/react-native-gesture-handler) + [react native svg](https://github.com/react-native-svg/react-native-svg). Reanimated and gesture handler expose declarative APIs in Javascript/Typescript, but don't run on the single JavaScript thread, so they don't block or get blocked by any other JS that needs to run. Yet you can still have a shared values for displaying something like a discreet value on a line graph while panning across it (at a full 60fps).

An example of this approach, from the great "Can It Be Done In React Native" project: https://www.youtube.com/watch?v=i20R4lwkJzw (source code: https://github.com/wcandillon/can-it-be-done-in-react-native/blob/master/season4/src/Rainbow/Graph.tsx)

**Comment by liamjones** [comment]
> I would also like this, and agree it might be worth looking at successful react native graph libraries! But I wouldn't suggest victory-native as a good modern example. 😞 There are lots of known performance bottlenecks (re-renders and frame drops; 'touch -> display overlays' and 'panning -> display discreet values' are notoriously difficult to get working in a performant way).

Ah, thanks for the info, I wasn't aware. I only mentioned them because of the fact that they're also a React graphing library that has got an RN implementation.

**Comment by esthor** [comment]
Ofc, no problem. I'm really glad you kept this thread going and brought something to the discussion! It would be so valuable for the React community as a whole to have a modern, performant graph library working universally across all "in-tree" React platforms (web, ios, android)...a kid can dream, right?

**Comment by liamjones** [comment]
Would still love to have RN support in Nivo

**Comment by liamjones** [comment]
bump, would still love to see this at some point

**Comment by Parz1val1** [comment]
Would be awesome to be able to use Nivo charts in React Native, there's nothing comparable for React native so you'd be opening a big market!

**Comment by carsonDentinger** [comment]
Bumping this to say it would be great to see react native support for nivo 👀

**Comment by esthor** [comment]
Are there any things specifically folks are wanting in a react-native chart library? Is it just "nivo, but for react native" or are there specifics? A full port would probably be a huge lift, especially if you're looking 1:1 on the entire API surface.

Perhaps "nivo, but works with react-native" is not practical (or going to be prioritized, given this issue is 3.5 years old). So, maybe we can look at what folks actually want and start building something from there if there isn't a solution out there already?

Some thoughts on a modern react-native charting architecture:
- Not pure JS runtime (native or its own thread?)
- non-blocking execution
- non-blocking gestures
- probably [Reanimated](https://github.com/software-mansion/react-native-reanimated) for the animations
- probably [Gesture Handler](https://github.com/software-mansion/react-native-gesture-handler) for the gestures
- maybe [React Native Skia](https://github.com/Shopify/react-native-skia) for the layout & rendering?
- Fabric support

For basic linear charts in react native, these are a couple modern options today:
- [React Native Graph](https://github.com/margelo/react-native-graph) -- uses Reanimated, Gesture Handler and Skia; just line chart right now
- [React Native Wagmi Charts](https://github.com/coinjar/react-native-wagmi-charts) -- uses Reanimated, Gesture Handler, and SVG; just a line and candlestick chart right now
- [React Native Animated Charts](https://github.com/rainbow-me/react-native-animated-charts) -- uses Reanimated, Gesture Handler, SVG, Haptics; just a line chart & might not be actively maintained since October 2021

**Comment by esthor** [comment]
...it's might also be less difficult to go from a react-native chart library -> react web chart library (utilizing [react-native-web](https://github.com/necolas/react-native-web)) if you want feature parity in the same lib

**Comment by minseokk** [comment]
bump

**Comment by BoavistaLudwig** [comment]
bump

**Comment by BoavistaLudwig** [comment]
is there a way to crowd-source the development of it?

**Comment by ethansoup** [comment]
RN support would be amazing, no good libraries in RN that compare even closely to this one.

**Comment by udaysagartammina** [comment]
I really like the visualisations of this library. We have a product that has both a web application using react.js and iOS and android application using react-native. We would like to use a library that can provide visualisations with a similar look and feel for both the applications. react-native support would be really good. 

Thanks in advance.

**Comment by krupav03** [comment]
It would be great if we had react native support .

**Comment by YahyaBagia** [comment]
Would love react-native support

**Comment by goblin-ihx** [comment]
bump

**Comment by jakehartigan** [comment]
Would love even a so-so implementation for RN.

Something as simple as an SVG generated that we can send once a week as a report would be enough for now.

**Comment by YahyaBagia** [comment]
bump

**Comment by shalawfatah** [comment]
RN implementation would be great.

**Comment by Polo4444** [comment]
bump

**Comment by esthor** [comment]
This looks very promising -- https://github.com/FormidableLabs/victory-native-xl

TL;DR -- Modern architectural choices (Skia, Reanimated, Gesture Handler) for high performance and fine control. (Re)Built by the group that made one of the most popular react native charting libraries for years.

**Comment by Soderhamn** [comment]
Bump

**Comment by StevenHarlow** [comment]
Would also love RN support

**Comment by jakehartigan** [comment]
Bump

**Comment by LeoAnt02** [comment]
bump

**Comment by Rayne182** [comment]
Agree, +1

**Comment by anzhou99** [comment]
bump

**Comment by ax-at** [comment]
Will react-native support for server components and the fact that nivo supports server-side rendering automatically make it compatible with react-native?

**Comment by esthor** [comment]
> Will react-native support for server components and the fact that nivo supports server-side rendering automatically make it compatible with react-native?

**Short answer**: No.

**Longer answer**: For something to be "compatible" with React Native, it typically means that either the implementation is target platform agnostic (like some "pure" libraries) or more commonly, it has native implementations (either using the built-in react native primitives or with its own native bindings, at least for iOS and Android target platforms).

But, Nivo uses web platform primitives (e.g. HTML tags, CSS properties, etc.), even in [`core`](https://github.com/plouc/nivo/tree/master/packages/core).

So, to make it "react native compatible", it's not enough to throw a web implementation into a webview in a native mobile app and call it a day. Basically, someone awesome, or a group of awesome someones, would need to reimplement the platform-specifics of this library. One way might be to create a platform-agnostic interface for everything (so any platform target bindings could be written more easily in the future, assuming it'd still be React). Another approach might be to just write the react native specific implementations as additive or perhaps even as transformation functions that know how refactor the web specifics to react native abstractions...just spit-balling here. The main point is, it'd be a lot of work.

**YOLO answer**: That said, if you want it to just be in a mobile app, and you have it working in a mobile-responsive way on web, and you don't really care that much about the user experience in a native mobile app, there's probably nothing technically that's ever stopped anyone from throwing it in a [react native web view](https://github.com/react-native-webview/react-native-webview) and YOLOing it.

I hope that helps. 🙏

**Comment by mohdfaizan5** [comment]
Bump++
